### ISSUE TITLE
🔥iOS bug: NSGenericException: Collection was mutated while being enumerated

### ISSUE THREAD

**Issue by vmaykovsky** [invertase/react-native-firebase#2180]
## Issue

Hello guys, I'm getting iOS app crash using **two Firebase projects in the same app** when reloading the app in emulator using `⌘+R`. Both Firebase projects are initialized in native code (see my `AppDelegate.m` below).

Fatal Exception at **`RNFirebaseAuth.m - Line 68`** (react-native-firebase v5.4.0):
 `*** Terminating app due to uncaught exception 'NSGenericException', reason: '*** Collection <__NSDictionaryM: 0x600001fb2b80> was mutated while being enumerated.'`

It's pretty obvious that this exception will occur in the following two functions in `RNFirebaseAuth.m` (lines 49-79) in case if you use two or more Firebase projects in your app:
```objc
- (void)dealloc {
  DLog(@"RNFirebaseAuth:instance-destroyed");
  
  for(NSString* key in authStateHandlers) {
    FIRApp *firApp = [RNFirebaseUtil getApp:key];
    [[FIRAuth authWithApp:firApp] removeAuthStateDidChangeListener:[authStateHandlers valueForKey:key]];
    [authStateHandlers removeObjectForKey:key];
  }
  
  for(NSString* key in idTokenHandlers) {
    FIRApp *firApp = [RNFirebaseUtil getApp:key];
    [[FIRAuth authWithApp:firApp] removeIDTokenDidChangeListener:[idTokenHandlers valueForKey:key]];
    [idTokenHandlers removeObjectForKey:key];
  }
}

- (void)invalidate {
  // dealloc sometimes is not called when app is reloaded.

  for(NSString* key in authStateHandlers) {
      FIRApp *firApp = [RNFirebaseUtil getApp:key];
      [[FIRAuth authWithApp:firApp] removeAuthStateDidChangeListener:[authStateHandlers valueForKey:key]];
      [authStateHandlers removeObjectForKey:key];
  }
  
  for(NSString* key in idTokenHandlers) {
      FIRApp *firApp = [RNFirebaseUtil getApp:key];
      [[FIRAuth authWithApp:firApp] removeIDTokenDidChangeListener:[idTokenHandlers valueForKey:key]];
      [idTokenHandlers removeObjectForKey:key];
  }
}
```
**All the `for` statements have to be replaced with a correct code.**

To reproduce the issue it's enough to initialize `Authentification` for both FB projects somewhere in your React Native code:
```js
const userDefault = firebase.auth().currentUser;
const userMaster = firebase.app('MASTER').auth().currentUser;
```
---

### iOS

<details><summary>Click To Expand</summary>
<p>

#### `ios/Podfile`:

- [ ] I'm not using Pods
- [x] I'm using Pods and my Podfile looks like:

```ruby
# Uncomment the next line to define a global platform for your project
platform :ios, '9.0'

target 'TheProject' do
  # Uncomment the next line if you're using Swift or would like to use dynamic frameworks
  # use_frameworks!

  # Pods for TheProject
  pod 'Firebase/Core', '~> 5.20.1'
  pod 'Firebase/Auth', '~> 5.20.1'
  pod 'Firebase/Firestore', '~> 5.20.1'
  pod 'Firebase/Messaging', '~> 5.20.1'
  pod 'Firebase/RemoteConfig', '~> 5.20.1'
  pod 'GoogleIDFASupport', '~> 3.14.0' 
  pod 'Fabric', '~> 1.9.0'
  pod 'Crashlytics', '~> 3.12.0'

end
```

#### `AppDelegate.m`:

```objc
#import <Firebase.h>
#import "RNFirebaseNotifications.h"
#import "RNFirebaseMessaging.h"

#import "AppDelegate.h"
#import <CodePush/CodePush.h>

#import <React/RCTBridge.h>
#import <React/RCTBundleURLProvider.h>
#import <React/RCTRootView.h>
#import "RNSplashScreen.h"

@implementation AppDelegate

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
  // Configure default Firebase (GoogleService-Info.plist)
  [FIRApp configure];

  // Configure second instance with manual options.
  FIROptions *firOptions = [[FIROptions alloc] initWithGoogleAppID:@"<GoogleAppID>" GCMSenderID:@"<GCMSenderID>"];
  firOptions.bundleID = @"<bundleID>";
  firOptions.APIKey = @"<APIKey>";
  firOptions.clientID = @"<clientID>";
  firOptions.databaseURL = @"<databaseURL>";
  firOptions.storageBucket = @"<storageBucket>";
  firOptions.projectID = @"<projectID>";
  
  // Configure an alternative FIRApp.
  [FIRApp configureWithName:@"MASTER" options:firOptions];

  [RNFirebaseNotifications configure];

  RCTBridge *bridge = [[RCTBridge alloc] initWithDelegate:self launchOptions:launchOptions];
  RCTRootView *rootView = [[RCTRootView alloc] initWithBridge:bridge moduleName:@"TheProject" initialProperties:nil];

  rootView.backgroundColor = [[UIColor alloc] initWithRed:1.0f green:1.0f blue:1.0f alpha:1];

  self.window = [[UIWindow alloc] initWithFrame:[UIScreen mainScreen].bounds];
  UIViewController *rootViewController = [UIViewController new];
  rootViewController.view = rootView;
  self.window.rootViewController = rootViewController;
  [self.window makeKeyAndVisible];
  [RNSplashScreen showSplash:@"LaunchScreen" inRootView:rootView];
  return YES;
}

- (NSURL *)sourceURLForBridge:(RCTBridge *)bridge
{
#if DEBUG
  return [[RCTBundleURLProvider sharedSettings] jsBundleURLForBundleRoot:@"index" fallbackResource:nil];
#else
  return [CodePush bundleURL];
#endif
}

// Firebase notifications and messages
- (void)application:(UIApplication *)application didReceiveLocalNotification:(UILocalNotification *)notification {
  [[RNFirebaseNotifications instance] didReceiveLocalNotification:notification];
}

- (void)application:(UIApplication *)application didReceiveRemoteNotification:(nonnull NSDictionary *)userInfo
                                                       fetchCompletionHandler:(nonnull void (^)(UIBackgroundFetchResult))completionHandler{
  [[RNFirebaseNotifications instance] didReceiveRemoteNotification:userInfo fetchCompletionHandler:completionHandler];
}

- (void)application:(UIApplication *)application didRegisterUserNotificationSettings:(UIUserNotificationSettings *)notificationSettings {
  [[RNFirebaseMessaging instance] didRegisterUserNotificationSettings:notificationSettings];
}

@end

```

</p>
</details>

---

## Environment

<details><summary>Click To Expand</summary>
<p>

**`react-native info` output:**

```
  React Native Environment Info:
    System:
      OS: macOS 10.14.5
      CPU: (8) x64 Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
      Memory: 72.32 MB / 16.00 GB
      Shell: 3.2.57 - /bin/bash
    Binaries:
      Node: 10.15.3 - ~/.nvm/versions/node/v10.15.3/bin/node
      Yarn: 1.16.0 - /usr/local/bin/yarn
      npm: 6.4.1 - ~/.nvm/versions/node/v10.15.3/bin/npm
      Watchman: 4.9.0 - /usr/local/bin/watchman
    SDKs:
      iOS SDK:
        Platforms: iOS 12.2, macOS 10.14, tvOS 12.2, watchOS 5.2
      Android SDK:
        API Levels: 23, 24, 25, 26, 27, 28
        Build Tools: 26.0.2, 28.0.3, 29.0.0
        System Images: android-16 | Google APIs Intel x86 Atom, android-23 | Google APIs ARM EABI v7a, android-23 | Google APIs Intel x86 Atom, android-28 | Google Play Intel x86 Atom
    IDEs:
      Android Studio: 3.4 AI-183.6156.11.34.5522156
      Xcode: 10.2.1/10E1001 - /usr/bin/xcodebuild
    npmPackages:
      react: 16.8.3 => 16.8.3 
      react-native: 0.59.8 => 0.59.8 
    npmGlobalPackages:
      react-native-asset: 1.1.3
      react-native-cli: 2.0.1
```

- **Platform that you're experiencing the issue on**:
  - [x] iOS

- **`Firebase` module(s) you're using that has the issue:**
  - `Authentication`

- **Are you using `TypeScript`?**
  - `Y`
  
</p>
</details>

**Comment by vmaykovsky** [comment]
Guys, could you include a fix for the issue into next release? It seems like a pretty quick and simple fix.

**Comment by mikehardy** [comment]
You could be correct with your fix diagnosis but I haven't seen this before and have no reproduction case for it. I did however just do a mountain of work so that v5 is testable, is releasable, and now has some releases (awesome, right?), so you have a chance to get the fix in, yes

Please propose a PR that passes the ios e2e tests (the tests README.md is complete and correct - I just worked through it over last 2 weeks) and if it looks good I'll include it in the next patch release

As a bonus, after working up the fix, you can include it via patch-package until it's in mainline

**Comment by vmaykovsky** [comment]
It would be great if you merge my pull request into next 5.x.x release

**Comment by mikehardy** [comment]
I linked the related PR and it was just blocked by a CI thing. We’re investigating that and we’ll see - should be possible?

**Comment by vmaykovsky** [comment]
Great, I see that ci.android.test failed, which is weird because it was just a small iOS fix.

**Comment by Salakar** [comment]
Sorting the CI out now sorry, have triggered a re-run, should hopefully be ok now 🙈

**Comment by mikehardy** [comment]
@Salakar if you like the bugfix then please approve, my iOS skills are still sad ;-). I have a typescript change queued up as well for a 5.5.5 - both changes are small so I am still comfortable with doing a release even though I travel soon

**Comment by Salakar** [comment]
👋 this is fixed in v5.5.5 - [[changelog]](https://rnfirebase.io/docs/v5.x.x/releases/v5.5.x#Changelog)
